# Simple Exploration of WOT Data


- env: Python 2.7 (`graphlab` library can only run in Python 2.7.x env)


- All statistical pics are available in two forms: screenshots & inline canvas (displays may abnormal after posting to Internet)
    - Vis all drawn on `GraphLab Canvas` (an interactive vis platform) in another browser tab by default
    
        - I didn't use `matplotlib` or `seaborn` for they were failing to tackle the drawing task because of large data volume
        
    - Vis can also display in Notebook but need to have `graphlab` and `jupyter notebook` installed first


- Chen, Jan. 11, 2020

In [1]:
import os
import time
import graphlab as gl

gl.canvas.set_target('ipynb')
# reset
# gl.canvas.set_target('browser')

In [2]:
data_folder = './data/'
source = os.listdir(data_folder)
for idx, filename in enumerate(source):
    print('fileID: %s, fileName: %s, fileSize: %s M'%(idx, filename, os.path.getsize(data_folder + filename) / float(1024 ** 2)))

fileID: 0, fileName: table_2.csv, fileSize: 40.4553222656 M
fileID: 1, fileName: table_3.csv, fileSize: 274.100172043 M
fileID: 2, fileName: table_1.tsv, fileSize: 8713.45498848 M


## A glimpse to `table_1.tsv` 

>battles on a particular day

In [3]:
%%time
# if using Pandas, importing data would be a mission impossible
battle = gl.SFrame.read_csv(data_folder + source[2], 
                            header=False, 
                            delimiter='\t')

This non-commercial license of GraphLab Create for academic use is assigned to luoc18@mails.tsinghua.edu.cn and will expire on December 19, 2020.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1578853743.log


Finished parsing file /Users/lyndon/AnacondaProjects/wangliao/WOT/data/table_1.tsv

Parsing completed. Parsed 100 lines in 0.790334 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,int,int,int,int,int,int,int,int,int,str,int,int,str,str,int,int,int,int,str,int,int,int,int,int,int,int,str,str,str,str,int,int,int,int,int]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 246979 lines. Lines per second: 202536

Read 1728833 lines. Lines per second: 256802

Read 2963830 lines. Lines per second: 253775

Read 4445771 lines. Lines per second: 250836

Read 5680699 lines. Lines per second: 248956

Read 7162668 lines. Lines per second: 249922

Read 8644749 lines. Lines per second: 251065

Read 9879772 lines. Lines per second: 246649

Read 11114719 lines. Lines per second: 245489

Read 12349636 lines. Lines per second: 245576

Read 13584644 lines. Lines per second: 242878

Read 14819656 lines. Lines per second: 241488

Read 16054567 lines. Lines per second: 241054

Read 17289511 lines. Lines per second: 241169

Read 18524527 lines. Lines per second: 240355

Read 19759460 lines. Lines per second: 240020

Read 20994444 lines. Lines per second: 240108

Read 22229490 lines. Lines per second: 239476

Read 23464552 lines. Lines per second: 239307

Read 24699480 lines. Lines per second: 239574

Read 25934470 lines. Lines per second: 238802

Read 27169404 lines. Lines per second: 237690

Read 28404392 lines. Lines per second: 237390

Read 29639387 lines. Lines per second: 237845

Read 30874327 lines. Lines per second: 237326

Read 32109255 lines. Lines per second: 237109

Read 33344268 lines. Lines per second: 237203

Read 34579345 lines. Lines per second: 236906

Read 35814372 lines. Lines per second: 236094

Read 36802423 lines. Lines per second: 234660

Read 38037359 lines. Lines per second: 233184

Read 39025386 lines. Lines per second: 231818

Read 40260335 lines. Lines per second: 231628

Read 41495366 lines. Lines per second: 231197

Read 42730375 lines. Lines per second: 230026

Finished parsing file /Users/lyndon/AnacondaProjects/wangliao/WOT/data/table_1.tsv

Parsing completed. Parsed 43043874 lines in 187.219 secs.

CPU times: user 8min 54s, sys: 18.7 s, total: 9min 13s
Wall time: 3min 22s


In [4]:
battle

Columns:
	X1	str
	X2	int
	X3	int
	X4	int
	X5	int
	X6	int
	X7	int
	X8	int
	X9	int
	X10	int
	X11	str
	X12	int
	X13	int
	X14	str
	X15	str
	X16	int
	X17	int
	X18	int
	X19	int
	X20	str
	X21	int
	X22	int
	X23	int
	X24	int
	X25	int
	X26	int
	X27	int
	X28	str
	X29	str
	X30	str
	X31	str
	X32	int
	X33	int
	X34	int
	X35	int
	X36	int

Rows: 43043874

Data:
+------------+------------+------+-------+----+------------+------+
|     X1     |     X2     |  X3  |   X4  | X5 |     X6     |  X7  |
+------------+------------+------+-------+----+------------+------+
| 2018-03-02 | 1001688791 | 5371 | 86441 | 1  |     0      |  0   |
| 2018-03-04 | 1030117817 |  0   |   84  | 1  |     0      |  0   |
| 2018-03-17 | 1011888852 | 7690 | 11089 | 2  | 1000011646 | 215  |
| 2018-03-19 | 1025052088 | 4291 |  5409 | 2  |     0      |  0   |
| 2018-03-20 | 1022562093 | 2463 |  2829 | 2  | 1000051254 |  15  |
| 2018-03-20 | 1022562093 | 2463 |  2829 | 2  | 1000051254 |  15  |
| 2018-03-09 | 1005199243 | 5484 | 25591 | 1  | 1000010436 | 1522 |
| 2018-03-29 | 1017641348 | 3713 | 40054 | 1  | 1000023668 | 619  |
| 2018-03-22 | 1022889983 | 1756 |  830  | 1  |     0      |  0   |
| 2018-03-05 | 1000519973 | 3049 |  2435 | 2  |     0      |  0   |
+------------+------------+------+-------+----+------------+------+
+-------------------+----------+-----+-------+-----+-------+------------+
|         X8        |    X9    | X10 |  X11  | X12 |  X13  |    X14     |
+-------------------+----------+-----+-------+-----+-------+------------+
|  7971503771003267 |    0     |  0  |  Dead | 107 |  9233 |    SPG     |
| 41532455531610259 |    0     |  0  |  Dead | 304 |  5665 | mediumTank |
| 34268128236899513 |    0     |  0  |  Dead | 160 |  1057 | mediumTank |
| 55043056846447463 | 14772806 |  0  |  Dead | 117 | 10497 | heavyTank  |
| 11046816320630794 | 3955515  |  0  |  Dead | 219 | 10497 | heavyTank  |
| 11046816320630794 | 3955515  |  0  |  Dead | 219 | 10497 | heavyTank  |
| 17850993704353163 |    0     |  0  |  Dead | 156 | 55569 |   AT-SPG   |
|  5906535037059825 |    0     |  0  | Alive | 276 |  2817 | heavyTank  |
|  5907845001494237 |    0     |  0  |  Dead | 399 |  5649 |    SPG     |
| 40675003965684398 |    0     |  0  |  Dead | 242 | 11777 | heavyTank  |
+-------------------+----------+-----+-------+-----+-------+------------+
+--------------------+-----+------+-----+-----+-----+-----+
|        X15         | X16 | X17  | X18 | X19 | X20 | ... |
+--------------------+-----+------+-----+-----+-----+-----+
|  germany:G61_G_E   |  10 | 338  | 493 | 550 |  0  | ... |
|   usa:A24_T2_med   |  2  |  0   |  0  |  0  |  \N | ... |
| usa:A05_M4_Sherman |  5  | 956  | 880 | 460 |  0  | ... |
|    ussr:R77_KV2    |  6  | 666  | 336 | 860 |  0  | ... |
|    ussr:R77_KV2    |  6  | 772  | 210 | 810 |  0  | ... |
|    ussr:R77_KV2    |  6  | 772  | 210 | 810 |  0  | ... |
|  germany:G48_E-25  |  7  | 779  | 287 | 830 |  0  | ... |
|   ussr:R106_KV85   |  6  | 1234 | 262 | 572 |  0  | ... |
| germany:G23_Grille |  5  | 674  |  0  | 240 |  0  | ... |
|    ussr:R80_KV1    |  5  | 960  |  0  | 590 |  0  | ... |
+--------------------+-----+------+-----+-----+-----+-----+
[43043874 rows x 36 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

> `table_1.tsv` contains 43,043,874 rows * 36 columns 

### How many users?

> 279,313 users

In [5]:
# X2 = user identifier (SPA ID)
battle['X2'] = battle['X2'].astype(str)
battle['X2'].sketch_summary()


+------------------+----------+----------+
|       item       |  value   | is exact |
+------------------+----------+----------+
|      Length      | 43043874 |   Yes    |
| # Missing Values |    0     |   Yes    |
| # unique values  |  280447  |    No    |
+------------------+----------+----------+

Most frequent items:
+-------+------------+------------+------------+------------+------------+------------+
| value | 1013840838 | 1029902210 | 1002992417 | 1024651570 | 1029856499 | 1009020303 |
+-------+------------+------------+------------+------------+------------+------------+
| count |    6319    |    6113    |    5470    |    5444    |    5426    |    5417    |
+-------+------------+------------+------------+------------+------------+------------+
+------------+------------+------------+------------+
| 1021168662 | 1018414070 | 1030058441 | 1004056671 |
+------------+------------+------------+------------+
|    5112    |    4530    |    4490    |    4388    |
+------------+------

In [6]:
%%time
# double check
print(len(set(list(battle['X2']))))

279313
CPU times: user 38.8 s, sys: 2.04 s, total: 40.9 s
Wall time: 40.8 s


### How many battles?

> 1,825,553 battles

In [7]:
# X8 = battle identifier
battle['X8'] = battle['X8'].astype(str)
battle['X8'].sketch_summary()


+------------------+----------+----------+
|       item       |  value   | is exact |
+------------------+----------+----------+
|      Length      | 43043874 |   Yes    |
| # Missing Values |    0     |   Yes    |
| # unique values  | 1827786  |    No    |
+------------------+----------+----------+

Most frequent items:
 -- All elements appear with less than 0.01% frequency -- 


In [8]:
%%time
# double check
print(len(set(list(battle['X8']))))

1825553
CPU times: user 49.2 s, sys: 2.78 s, total: 52 s
Wall time: 52.2 s


### Amount of time user was in the battle total

In [9]:
# X12 = amount of time user was in the battle total
battle['X12'].show(view='Numeric')

In [10]:
%%html
<img src = './imgs/X12.png', width=1400, height=600>

### Battle start time

> earliest start time: `2018-03-01 00:00:00`

In [11]:
%%time
# X30 = battle start time
battle['stTimestamp'] = battle['X30'].apply(lambda x: time.mktime(time.strptime(x, "%Y-%m-%d %H:%M:%S")))

CPU times: user 51.4 ms, sys: 207 ms, total: 259 ms
Wall time: 8.68 s


In [12]:
battle['stTimestamp'].sketch_summary()


+--------------------+-------------------+----------+
|        item        |       value       | is exact |
+--------------------+-------------------+----------+
|       Length       |      43043874     |   Yes    |
|        Min         |    1519891200.0   |   Yes    |
|        Max         |    1522565999.0   |   Yes    |
|        Mean        |   1521273188.26   |   Yes    |
|        Sum         | 6.54814914351e+16 |   Yes    |
|      Variance      | 6.12386349619e+11 |   Yes    |
| Standard Deviation |   782551.180192   |   Yes    |
|  # Missing Values  |         0         |   Yes    |
|  # unique values   |      1274735      |    No    |
+--------------------+-------------------+----------+

Most frequent items:
 -- All elements appear with less than 0.01% frequency -- 

Quantiles: 
+--------------+--------------+--------------+--------------+--------------+
|      0%      |      1%      |      5%      |     25%      |     50%      |
+--------------+--------------+--------------+---

In [13]:
# calculate the exact start time
print(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(1519891200.0)))

2018-03-01 00:00:00


In [14]:
battle['stInterval'] = battle['stTimestamp'] - 1519891200.0
battle['stInterval'].show(view='Numeric')

In [15]:
%%html
<img src='./imgs/X30.png', width=1400, height=600>

### Battle time duration

In [16]:
# X32 = battle time duration
battle['X32'].sketch_summary()


+--------------------+---------------+----------+
|        item        |     value     | is exact |
+--------------------+---------------+----------+
|       Length       |    43043874   |   Yes    |
|        Min         |      0.0      |   Yes    |
|        Max         |     1800.0    |   Yes    |
|        Mean        | 382.422491479 |   Yes    |
|        Sum         | 16460945538.0 |   Yes    |
|      Variance      | 18105.6713565 |   Yes    |
| Standard Deviation |  134.55731625 |   Yes    |
|  # Missing Values  |       0       |   Yes    |
|  # unique values   |      1449     |    No    |
+--------------------+---------------+----------+

Most frequent items:
+-------+--------+--------+--------+--------+--------+--------+--------+--------+
| value |  600   |  900   |  331   |  322   |  321   |  311   |  305   |  330   |
+-------+--------+--------+--------+--------+--------+--------+--------+--------+
| count | 321960 | 298746 | 178699 | 178473 | 176620 | 176603 | 176569 | 176483 |

In [17]:
battle['X32'].show(view='Numeric')

In [18]:
%%html
<img src='./imgs/X32.png', width=1400, height=600>

### Team textual char var: # of chat entries in battle

In [19]:
# X33 = # of chat entries in battle
battle['X33'].sketch_summary()


+--------------------+---------------+----------+
|        item        |     value     | is exact |
+--------------------+---------------+----------+
|       Length       |    43043874   |   Yes    |
|        Min         |      0.0      |   Yes    |
|        Max         |     358.0     |   Yes    |
|        Mean        | 4.04948344101 |   Yes    |
|        Sum         |  174305455.0  |   Yes    |
|      Variance      | 118.142038547 |   Yes    |
| Standard Deviation |  10.869316379 |   Yes    |
|  # Missing Values  |       0       |   Yes    |
|  # unique values   |      269      |    No    |
+--------------------+---------------+----------+

Most frequent items:
+-------+----------+--------+--------+--------+--------+--------+--------+
| value |    0     |   10   |   11   |   9    |   12   |   8    |   13   |
+-------+----------+--------+--------+--------+--------+--------+--------+
| count | 34186393 | 383940 | 383773 | 380790 | 374385 | 366835 | 361465 |
+-------+----------+-------

In [20]:
battle['X33'].show(view='Numeric')

In [21]:
%%html
<img src='./imgs/X33.png', width=1400, height=600>

## A glimpse to `table_3.csv`

> user table

In [22]:
%%time
user = gl.SFrame.read_csv(data_folder + source[1], 
                          header=True, 
                          column_type_hints={'battle_cnt_sandbox': int})

Finished parsing file /Users/lyndon/AnacondaProjects/wangliao/WOT/data/table_3.csv

Parsing completed. Parsed 100 lines in 0.965351 secs.

Read 792355 lines. Lines per second: 591844

Finished parsing file /Users/lyndon/AnacondaProjects/wangliao/WOT/data/table_3.csv

Parsing completed. Parsed 4344092 lines in 4.81848 secs.

CPU times: user 21 s, sys: 913 ms, total: 21.9 s
Wall time: 7.38 s


In [23]:
user

Columns:
	spa_key	int
	first_battle_dt	str
	country_name	str
	lang	str
	battle_cnt_sandbox	int
	bttl_training_cnt_in_first_month	int
	purchase	float
	active	str

Rows: 4344092

Data:
+------------+---------------------+---------------+------+--------------------+
|  spa_key   |   first_battle_dt   |  country_name | lang | battle_cnt_sandbox |
+------------+---------------------+---------------+------+--------------------+
| 1030092313 | 2018-03-03 19:54:44 |    Ukraine    |  ru  |         0          |
| 1001054194 | 2011-07-24 08:31:35 |   Australia   |  en  |         0          |
| 1004501051 | 2013-04-10 23:59:59 | United States |  en  |         0          |
| 1002037736 | 2012-03-10 23:59:59 | United States |  en  |         0          |
| 1002678822 | 2012-07-09 23:59:59 | United States |  en  |         0          |
| 1000446568 | 2013-08-09 23:59:59 | United States |  en  |         0          |
| 1020071257 | 2016-04-23 18:22:51 |     Canada    |  en  |         0          |
| 1017894257 | 2016-01-03 15:52:12 |     Norway    |  en  |         0          |
| 1022808564 | 2016-10-28 20:00:04 |    Ukraine    |  en  |         0          |
| 1009757176 | 2015-06-27 12:31:50 | United States |  en  |         0          |
+------------+---------------------+---------------+------+--------------------+
+-------------------------------+----------+------------+
| bttl_training_cnt_in_first... | purchase |   active   |
+-------------------------------+----------+------------+
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
|               0               |   0.0    | not active |
+-------------------------------+----------+------------+
[4344092 rows x 8 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### User's battle count in sandbox matches

In [24]:
user['battle_cnt_sandbox'].sketch_summary()


+--------------------+------------------+----------+
|        item        |      value       | is exact |
+--------------------+------------------+----------+
|       Length       |     4344092      |   Yes    |
|        Min         |       0.0        |   Yes    |
|        Max         |       67.0       |   Yes    |
|        Mean        | 0.00450542944302 |   Yes    |
|        Sum         |     19572.0      |   Yes    |
|      Variance      | 0.0373260556487  |   Yes    |
| Standard Deviation |   0.1931995229   |   Yes    |
|  # Missing Values  |        0         |   Yes    |
|  # unique values   |        42        |    No    |
+--------------------+------------------+----------+

Most frequent items:
+-------+---------+------+------+-----+-----+-----+-----+-----+-----+-----+
| value |    0    |  1   |  2   |  3  |  4  |  5  |  10 |  6  |  7  |  9  |
+-------+---------+------+------+-----+-----+-----+-----+-----+-----+-----+
| count | 4337624 | 3287 | 1079 | 574 | 365 | 227 | 222 | 16

In [25]:
user['battle_cnt_sandbox'].show(view='Numeric')

In [26]:
%%html
<img src='./imgs/user_battleCountSandbox.png', width=1400, height=600>